core/src/main/java/org/apache/iceberg/actions/RewriteFileGroupPlanner.java

```diff
@@ -0,0 +1,175 @@
+
+  private final FileRewriter<FileScanTask, DataFile> rewriter;
+  private final RewriteJobOrder rewriteJobOrder;
+
+  public RewriteFileGroupPlanner(
+      FileRewriter<FileScanTask, DataFile> rewriter, RewriteJobOrder rewriteJobOrder) {
+    this.rewriter = rewriter;
+    this.rewriteJobOrder = rewriteJobOrder;
+  }
+
+  public RewritePlanResult plan(Table table, Expression filter, long startingSnapshotId) {
+    StructLikeMap<List<List<FileScanTask>>> plan =
+        planFileGroups(table, filter, startingSnapshotId);
+    RewriteExecutionContext ctx = new RewriteExecutionContext();
+    Stream<RewriteFileGroup> groups =
+        plan.entrySet().stream()
+            .filter(e -> !e.getValue().isEmpty())
+            .flatMap(
+                e -> {
+                  StructLike partition = e.getKey();
+                  List<List<FileScanTask>> scanGroups = e.getValue();
+                  return scanGroups.stream().map(tasks -> newRewriteGroup(ctx, partition, tasks));
+                })
+            .sorted(RewriteFileGroup.comparator(rewriteJobOrder));
+    Map<StructLike, Integer> groupsInPartition = plan.transformValues(List::size);
+    int totalGroupCount = groupsInPartition.values().stream().reduce(Integer::sum).orElse(0);
+    return new RewritePlanResult(groups, totalGroupCount, groupsInPartition);
+  }
+
+  private StructLikeMap<List<List<FileScanTask>>> planFileGroups(
+      Table table, Expression filter, long startingSnapshotId) {
+    CloseableIterable<FileScanTask> fileScanTasks =
+        table
+            .newScan()
+            .useSnapshot(startingSnapshotId)
+            .filter(filter)
+            .ignoreResiduals()
+            .planFiles();
+
+    try {
+      Types.StructType partitionType = table.spec().partitionType();
+      StructLikeMap<List<FileScanTask>> filesByPartition =
+          groupByPartition(table, partitionType, fileScanTasks);
+      return filesByPartition.transformValues(
+          tasks -> ImmutableList.copyOf(rewriter.planFileGroups(tasks)));
+    } finally {
+      try {
+        fileScanTasks.close();
+      } catch (IOException io) {
+        LOG.error("Cannot properly close file iterable while planning for rewrite", io);
+      }
+    }
+  }
+
+  private StructLikeMap<List<FileScanTask>> groupByPartition(
+      Table table, Types.StructType partitionType, Iterable<FileScanTask> tasks) {
+    StructLikeMap<List<FileScanTask>> filesByPartition = StructLikeMap.create(partitionType);
+    StructLike emptyStruct = GenericRecord.create(partitionType);
+
+    for (FileScanTask task : tasks) {
+      // If a task uses an incompatible partition spec the data inside could contain values
+      // which belong to multiple partitions in the current spec. Treating all such files as
+      // un-partitioned and grouping them together helps to minimize new files made.
+      StructLike taskPartition =
+          task.file().specId() == table.spec().specId() ? task.file().partition() : emptyStruct;
```

Comment:
Minor fix; we can drop this now and use EmptyStructLike.get()

Reply:
Do we want to make `EmptyStructLike`, and `EmptyStructLike.get` public?
Currently it is package private.

Reply:
Made `EmptyStructLike` public for now, but if you think this change doesn't worth it to make it public, I can revert this part.

Reply:
Got this exception:
```
TestRewriteDataFilesAction > testBinPackRewriterWithSpecificOutputSpec() > formatVersion = 3 FAILED
    java.lang.UnsupportedOperationException: Can't retrieve values from an empty struct
        at org.apache.iceberg.EmptyStructLike.get(EmptyStructLike.java:40)
        at org.apache.iceberg.types.JavaHashes$StructLikeHash.hash(JavaHashes.java:96)
        at org.apache.iceberg.types.JavaHashes$StructLikeHash.hash(JavaHashes.java:75)
        at org.apache.iceberg.util.StructLikeWrapper.hashCode(StructLikeWrapper.java:96)
        at java.base/java.util.HashMap.hash(HashMap.java:338)
```

So reverted the `EmptyStructLike` changes

Reply:
Blargh